lib/battle/CBattleInfoCallback.cpp

```diff
@@ -1673,7 +1674,7 @@ SpellID CBattleInfoCallback::getRandomBeneficialSpell(CRandomGenerator & rand, c
 				return !stack->canShoot();
 			});
 
-			if(!walker)
+			if(!walker || subject->hasBonusOfType(BonusType::SPELL_SCHOOL_IMMUNITY, SpellSchool::FIRE))
```

Comment:
This block is also used by Shield spell, which is earth spell

Reply:
Strange. Why this block is needed? Code actually checks whether spell can be casted on this target - see `canBeCast` block above.

Reply:
the block does not work - if you try to cast genie spell on ballista then half of the attempts fail

Reply:
Ah, we actually check whether spell can be cast *at all*, not whether spell can be cast on this target. Will fix.